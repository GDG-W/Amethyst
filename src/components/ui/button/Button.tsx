"use client";
import React from "react";
type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  children: React.ReactNode;
  loading?: boolean;
  variant?: "primary" | "secondary" | "link";
  size?: "large" | "normal" | "full";
};
const Button = ({ children, disabled, variant, size, loading = false, ...rest }: ButtonProps) => {
  const buttonStyles = {
    variants: {
      primary: "bg-away-base text-white ",
      secondary: "bg-away-base/10 border border-away-base text-away-base",
      link: "bg-transparent text-away-base underline underline-offset-2",
    },
    sizes: {
      large: "w-full sm:w-40",
      full: " w-full",
      normal: "w-full sm:w-fit",
    },
  };
  const buttonVariant = variant || "primary";
  const buttonSize = size || "normal";
  const isDisabled = loading || disabled;
  return (
    <button
      disabled={isDisabled}
      className={`flex items-center justify-center gap-2 px-3 py-3 min-w-30 rounded-full capitalize hover:opacity-80 disabled:bg-bg-soft-200 disabled:text-white disabled:cursor-not-allowed ${buttonStyles.variants[buttonVariant]} ${buttonStyles.sizes[buttonSize]}
      `}
      {...rest}
    >
      {loading ? (
        <span className='h-5 w-5 border-2 border-white border-t-transparent rounded-full animate-spin'></span>
      ) : (
        children
      )}
    </button>
  );
};

export default Button;
