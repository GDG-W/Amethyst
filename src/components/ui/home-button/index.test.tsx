import "@testing-library/jest-dom";

import { fireEvent, render, screen } from "@testing-library/react";

import Button from ".";

type ButtonVariant = "primary" | "secondary" | "ghost";

describe("Button Component", () => {
  describe("Rendering", () => {
    it("renders with children text", () => {
      render(<Button>Click me</Button>);
      expect(screen.getByRole("button")).toHaveTextContent("Click me");
    });

    it("renders with React node children", () => {
      render(
        <Button>
          <span data-testid="icon">🚀</span>
          <span>Launch</span>
        </Button>
      );
      expect(screen.getByTestId("icon")).toBeInTheDocument();
      expect(screen.getByRole("button")).toHaveTextContent("🚀Launch");
    });
  });

  describe("Variants", () => {
    it("renders primary variant by default", () => {
      render(<Button>Primary</Button>);
      const button = screen.getByRole("button");

      expect(button).toHaveClass("bg-[#F6B51E]");
      expect(button).toHaveClass("border-[#1E1E1E]");
      expect(button).toHaveClass("w-[10.9375rem]");
      expect(button).toHaveClass("md:w-[16.125rem]");
    });

    it("renders secondary variant correctly", () => {
      render(<Button variant="secondary">Secondary</Button>);
      const button = screen.getByRole("button");

      expect(button).toHaveClass("bg-[#1E1E1E]");
      expect(button).toHaveClass("border-2");
      expect(button).toHaveClass("border-white");
      expect(button).toHaveClass("text-white");
      expect(button).toHaveClass("w-[13.81rem]");
    });

    it("renders ghost variant correctly", () => {
      render(<Button variant="ghost">Ghost</Button>);
      const button = screen.getByRole("button");

      expect(button).toHaveClass("bg-[#1E1E1E]");
      expect(button).toHaveClass("border-none");
      expect(button).toHaveClass("text-white");
    });
  });

  describe("Base Classes", () => {
    it("applies all base classes", () => {
      render(<Button>Test</Button>);
      const button = screen.getByRole("button");

      expect(button).toHaveClass("p-4");
      expect(button).toHaveClass("h-[52px]");
      expect(button).toHaveClass("md:h-[3.75rem]");
      expect(button).toHaveClass("rounded-full");
      expect(button).toHaveClass("heading-6");
      expect(button).toHaveClass("border-4");
      expect(button).toHaveClass("transition-all");
      expect(button).toHaveClass("duration-200");
      expect(button).toHaveClass("hover:scale-105");
      expect(button).toHaveClass("font-akira");
    });
  });

  describe("Click Handling", () => {
    it("calls onClick when provided", () => {
      const mockOnClick = jest.fn();
      render(<Button onClick={mockOnClick}>Clickable</Button>);

      fireEvent.click(screen.getByRole("button"));
      expect(mockOnClick).toHaveBeenCalledTimes(1);
    });

    it("does not throw error when onClick is not provided", () => {
      render(<Button>No Click Handler</Button>);

      expect(() => {
        fireEvent.click(screen.getByRole("button"));
      }).not.toThrow();
    });

    it("calls onClick multiple times when clicked multiple times", () => {
      const mockOnClick = jest.fn();
      render(<Button onClick={mockOnClick}>Multi Click</Button>);
      const button = screen.getByRole("button");

      fireEvent.click(button);
      fireEvent.click(button);
      fireEvent.click(button);

      expect(mockOnClick).toHaveBeenCalledTimes(3);
    });
  });

  describe("Custom Classes", () => {
    it("applies custom className", () => {
      render(<Button className="custom-class">Custom</Button>);
      const button = screen.getByRole("button");

      expect(button).toHaveClass("custom-class");
    });

    it("applies multiple custom classes", () => {
      render(<Button className="custom-1 custom-2">Multiple Custom</Button>);
      const button = screen.getByRole("button");

      expect(button).toHaveClass("custom-1");
      expect(button).toHaveClass("custom-2");
    });

    it("combines custom classes with base and variant classes", () => {
      render(
        <Button variant="secondary" className="extra-margin">
          Combined Classes
        </Button>
      );
      const button = screen.getByRole("button");

      expect(button).toHaveClass("p-4");
      expect(button).toHaveClass("rounded-full");

      expect(button).toHaveClass("bg-[#1E1E1E]");
      expect(button).toHaveClass("text-white");

      expect(button).toHaveClass("extra-margin");
    });

    it("handles empty className prop", () => {
      render(<Button className="">Empty Class</Button>);
      const button = screen.getByRole("button");

      expect(button).toHaveClass("p-4");
      expect(button).toHaveClass("bg-[#F6B51E]");
    });
  });

  describe("Integration Tests", () => {
    it("works with all props combined", () => {
      const mockOnClick = jest.fn();
      render(
        <Button variant="ghost" onClick={mockOnClick} className="integration-test">
          <span>🎯</span> Full Test
        </Button>
      );

      const button = screen.getByRole("button");

      expect(button).toHaveTextContent("🎯 Full Test");

      expect(button).toHaveClass("bg-[#1E1E1E]");
      expect(button).toHaveClass("border-none");
      expect(button).toHaveClass("p-4");
      expect(button).toHaveClass("integration-test");

      fireEvent.click(button);
      expect(mockOnClick).toHaveBeenCalledTimes(1);
    });

    it("maintains accessibility attributes", () => {
      render(<Button>Accessible Button</Button>);
      const button = screen.getByRole("button");

      expect(button.tagName).toBe("BUTTON");
      expect(button).toBeVisible();
      expect(button).toBeEnabled();
    });
  });

  describe("Edge Cases", () => {
    it("handles undefined variant gracefully", () => {
      render(<Button variant={undefined as unknown as ButtonVariant}>Undefined Variant</Button>);
      const button = screen.getByRole("button");

      expect(button).toHaveClass("bg-[#F6B51E]");
    });

    it("handles null children", () => {
      render(<Button>{null}</Button>);
      const button = screen.getByRole("button");

      expect(button).toBeInTheDocument();
      expect(button).toHaveTextContent("");
    });

    it("handles empty string children", () => {
      render(<Button>{""}</Button>);
      const button = screen.getByRole("button");

      expect(button).toBeInTheDocument();
      expect(button).toHaveTextContent("");
    });

    it("handles complex React node children", () => {
      render(
        <Button>
          <div>
            <span>Nested</span>
            <strong>Content</strong>
          </div>
        </Button>
      );

      const button = screen.getByRole("button");
      expect(button).toHaveTextContent("NestedContent");
      expect(button.querySelector("div")).toBeInTheDocument();
      expect(button.querySelector("span")).toBeInTheDocument();
      expect(button.querySelector("strong")).toBeInTheDocument();
    });
  });

  describe("Snapshots", () => {
    it("matches snapshot for primary variant", () => {
      const { container } = render(<Button>Primary Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });

    it("matches snapshot for secondary variant", () => {
      const { container } = render(<Button variant="secondary">Secondary Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });

    it("matches snapshot for ghost variant", () => {
      const { container } = render(<Button variant="ghost">Ghost Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
