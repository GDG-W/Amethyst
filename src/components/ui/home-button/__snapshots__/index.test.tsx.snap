// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Component Snapshots matches snapshot for ghost variant 1`] = `
<button
  class="p-4 h-[52px] md:h-[3.75rem] items-center justify-center flex  rounded-full heading-6 border-4 transition-all duration-200 hover:scale-105 font-akira bg-[#1E1E1E] border-none text-white  "
>
  Ghost Button
</button>
`;

exports[`Button Component Snapshots matches snapshot for primary variant 1`] = `
<button
  class="p-4 h-[52px] md:h-[3.75rem] items-center justify-center flex  rounded-full heading-6 border-4 transition-all duration-200 hover:scale-105 font-akira bg-[#F6B51E] border-[#1E1E1E] text-sm md:text-base w-[10.9375rem] md:w-[16.125rem]  "
>
  Primary Button
</button>
`;

exports[`Button Component Snapshots matches snapshot for secondary variant 1`] = `
<button
  class="p-4 h-[52px] md:h-[3.75rem] items-center justify-center flex  rounded-full heading-6 border-4 transition-all duration-200 hover:scale-105 font-akira bg-[#1E1E1E] border-2 border-white text-white w-[13.81rem] "
>
  Secondary Button
</button>
`;
